adds tests, refactors, comments

noahamac opened the pull request:
We've had nice adoption in open beta but are looking to gather more data with more users. 

[Daily active minutes by day](https://meta.looker.com/looks/50644)

[8500+ minutes used, 150 customers, 500+ users](https://meta.looker.com/explore/product_usage/page_events?qid=qijFLBNxMeUvfkQMYKxQdi&origin_space=2310&toggle=vis)


To do this we would like to move to Public Preview status on Wednesday. This PR adds tests, refactors, and comments to the project, responds to feedback from previous PRs, and initial feedback from open beta. 

Changes:
- [Restructured component files](https://drive.google.com/file/d/1NXZ757N_pq4Bjb7RA18hDqXnJLAUEWQI/view?usp=sharing&resourcekey=0-Enp5HQbK48HZaUVDyPjabA)
- More comments
- Unit & snapshot tests
- fixes click event bug on firefox (custy reported)
- [Improved distribution query](https://drive.google.com/file/d/1K21iNU7n0lCDUmHzJO9B72OpBbxT9795/view?usp=sharing&resourcekey=0-tZoeV2bxRUs9KbfW_HdLtg)
- Help section
![image](https://user-images.githubusercontent.com/56001784/110204474-0275d380-7e28-11eb-958a-e444577d6b2b.png)

**I am specifically looking for changes that would block moving to Public preview status.** I remain enthusiastically open to other, high level feedback, such as implementing redux and sagas, which I would like to do in a future update. The file system restructuring was great advice and provided an excellent refactoring opportunity.

Your feedback is welcome here, or on a doc, like Bryn's https://docs.google.com/document/d/1zyAGFHYST7oX8CHs9geyj-dOIgRVJX2Kp_zMLTetXfc/edit?usp=sharing . I will run prettier after this PR is approved

This build of the lookml diagram successfully diagrams all but 6 models on demoexpo which do not have connections configured and have lookml validation errors. Find it live on demoexpo: https://demoexpo.looker.com/extensions/lookml_model_diagram::dev_ext/models/thelooker-turtles/explores/order_items


Thank you very much for your time and wisdom :smiley:

RichardCzechowski commented on src/components/DiagramFrame/DiagramFrame.tsx:
```diff
@@ -0,0 +1,267 @@
+
+  function toggleExploreInfo() {
+    if (currentExplore && modelDetail) {
+      Object.keys(selectionInfo).length === 0 || selectionInfo.lookmlElement !== "explore"
+      ? setSelectionInfo({
+        lookmlElement: "explore"
+      })
+      : setSelectionInfo({})
+    }
```
Is this equivalent?  I'm not sure what other keys on `selectInfo` matter here, but I found the current implementation hard to read.

```suggestion
      setSelectionInfo({
        lookmlElement: selectionInfo.lookmlElement !== "explore" ? "explore" : undefined
      })
    }
```

RichardCzechowski commented on src/components/DiagramFrame/DiagramFrame.tsx:
```diff
@@ -0,0 +1,267 @@
+    }
+  }).sort((a: SelectOptionProps, b: SelectOptionProps) => a.label < b.label ? -1 : 1) : []
+
+  let currentExplore: ILookmlModelExplore = modelDetail && modelDetail.explores.filter((d: ILookmlModelExplore)=>{
+    return d.name === pathExploreName
+  })[0]
```
Optional chaining + implicit return cleans this up some. There are quite a few places in here you could do this if you were so inclined.

```suggestion
  let currentExplore: ILookmlModelExplore = modelDetail?.explores.filter((d: ILookmlModelExplore)=> d.name === pathExploreName)[0]
```

RichardCzechowski commented on src/components/DiagramFrame/DiagramFrame.tsx:
```diff
@@ -0,0 +1,267 @@
+    return d.name === pathExploreName
+  })[0]
+
+  let exploreList: ExploreDropdown[] = currentModel && currentModel.explores.map((d: ILookmlModelExplore)=>{
+    return {
+      value: d.name,
+      label: d.label
+    }
+  }).sort((a: ExploreDropdown, b: ExploreDropdown) => a.label < b.label ? -1 : 1)
```
This logic is repeated from above, you could factor out the map/sort into a util and use it both places.

RichardCzechowski commented on src/components/DiagramFrame/DiagramFrame.tsx:
```diff
@@ -0,0 +1,267 @@
+  let currentDiagramMetadata: DiagramMetadata = currentDimensions && currentDimensions.diagramDict
+
+  let defaultViews: VisibleViewLookup = {}
+  Object.keys(viewVisible).length === 0 && currentExplore && currentDiagramMetadata && Object.keys(currentDiagramMetadata.tableData)
+  .map((lookmlViewName: string)=>{
+    defaultViews[lookmlViewName] = true
+  })
```
Consider bringing in lodash for some nice utils. For example, `isEmpty` and `mapKeys` make this much nicer.

```suggestion
  if(isEmpty(viewVisible) && currentExplore && currentDiagramMetadata){ 
  mapKeys(currentDiagramMetadata.tableData, (lookmlViewName: string)=>{
    defaultViews[lookmlViewName] = true
  })
}
```

In general though, I suggest using if statements over conjunctions since they are more idiomatic

noahamac replied:
https://looker.atlassian.net/browse/MRKTI-470

Good call, have not used lodash before. Should implement in many places

RichardCzechowski commented on src/components/DiagramFrame/DiagramFrame.tsx:
```diff
@@ -0,0 +1,267 @@
+  .map((lookmlViewName: string)=>{
+    defaultViews[lookmlViewName] = true
+  })
+  Object.keys(viewVisible).length === 0 && currentExplore && currentDiagramMetadata && setViewVisible(defaultViews)
```
```suggestion
  if(isEmpty(viewVisible) && currentExplore && currentDiagramMetadata)
  {
    setViewVisible(defaultViews)
  }
```

RichardCzechowski commented on src/components/DiagramFrame/DiagramFrame.tsx:
```diff
@@ -0,0 +1,267 @@
+            label="Diagram"
+            tooltipPlacement="right"
+            size="large"
+            toggle={showNoAside && !showViewOptions && !showSettings && !showHelp ? true : undefined}
```
Another general point I won't comment on further, this kind of thing is much better in named functions. It's hard to understand a series of conditions like this, while a function could specifically tell me under what condition this toggle should occur. Even if it is nebulous like this, you can docstring it for the next person.

RichardCzechowski commented on src/components/DiagramFrame/MetadataPanel/MetadataPanel.tsx:
```diff
@@ -0,0 +1,196 @@
+import { METADATA_PANEL_PIXEL } from "../../../utils/constants"
+import { LookmlObjectMetadata, SelectionInfoPacket } from "../../interfaces"
+import { ILookmlModelExploreField, ILookmlModelExploreJoins } from '@looker/sdk/lib/sdk/3.1/models';
+import { getFields } from '../../../utils/diagrammer'
+import { exploreFieldURL } from '../../../utils/urls'
+import { useExplore } from '../../../utils/fetchers'
```
A nice pattern we use is to have an `index.ts` file in each folder that hoists the exports up to a single location for a clearer interface. Then you can have
`import { getFields, exploreFieldURL, useExplore } from '../../../utils` instead of three separate lines.
Same goest for your components etc. all throughout this PR

Also a better alternative to what you have below which is a single "utils" file with several utils in it. You might instead want a utils folder with a file called "getMetadata.ts" that exports those up to the utils index.ts file

noahamac replied:
https://looker.atlassian.net/browse/MRKTI-474

RichardCzechowski commented on src/components/DiagramFrame/MetadataPanel/MetadataPanel.tsx:
```diff
@@ -0,0 +1,196 @@
+}) => {
+  let metadata: LookmlObjectMetadata
+  let field: ILookmlModelExploreField
+  // @ts-ignore
```
Always nice to doc things like this. They can hide bugs.

RichardCzechowski commented on src/components/DiagramFrame/MetadataPanel/MetadataPanel.tsx:
```diff
@@ -0,0 +1,196 @@
+  let field: ILookmlModelExploreField
+  // @ts-ignore
+  let exploreLookmlLink = explore.lookml_link
+  if (selectionInfo.lookmlElement === "explore") {
```
You use the string "explore" all over (among others). Consider a constant

noahamac replied:
https://looker.atlassian.net/browse/MRKTI-473

RichardCzechowski commented on src/components/DiagramFrame/MetadataPanel/MetadataPanel.tsx:
```diff
@@ -0,0 +1,196 @@
+    })[0]
+    metadata = getJoinMetadata(joinObj, exploreLookmlLink)
+  } else if (selectionInfo.lookmlElement === "dimension" || selectionInfo.lookmlElement === "measure") {
+    let fields = getFields(explore.fields).filter((field: any) => {
```
I think the "any" here is a LookmlModelExploreField, do you have access to looker's types in here?

noahamac replied:
Yes, but for some reason `dimension_group` is not defined on the type definition for ILookmlModelExploreField ... I should file as a bug :slightly_smiling_face:

RichardCzechowski commented on src/components/DiagramFrame/MetadataPanel/utils.ts:
```diff
@@ -0,0 +1,111 @@
+  let sqlLine = join.sql_on && `  sql_on: ${join.sql_on} ;;\n`
+  let fkLine = join.foreign_key && `  foreign_key: ${join.foreign_key} \n`
+  let endLine = `}`
+  return [startLine, typeLine, relationLine, sqlLine, fkLine, endLine].filter(Boolean).join("")
```
clever

RichardCzechowski commented on src/components/DiagramFrame/types.ts:
```diff
@@ -0,0 +1,174 @@
+  font-style: italic;
+`
+
+export const SettingsPanel = styled(Aside as any)`
```
Mixing types and styled components again.

considering moving these to `DiagramFrame/components/SettingsPanel` etc

RichardCzechowski commented on src/components/DiagramFrame/types.ts:
```diff
@@ -0,0 +1,174 @@
+  margin: 0;
+`
+
+
```
White space

RichardCzechowski commented on src/components/DiagramFrame/types.ts:
```diff
@@ -0,0 +1,174 @@
+
+
+  `
+  
```
Lots of whitespace

RichardCzechowski commented on src/components/DiagramFrame/types.ts:
```diff
@@ -0,0 +1,174 @@
+  }
+
+
+`
```
Looks like this backtick is running away

RichardCzechowski commented on src/d3-utils/join-helpers.ts:
```diff
@@ -0,0 +1,55 @@
+import { 
+  TABLE_ROW_HEIGHT,
+} from '../utils/constants'
+
```
Docstrings in this file would be great (and more generally for all functions, exported components, etc)

noahamac replied:
https://looker.atlassian.net/browse/MRKTI-473

RichardCzechowski commented on src/utils/diagrammer.ts:
```diff
@@ -347,47 +361,55 @@ export function getDiagramDict(exploreFields: ILookmlModelExploreFieldset, joins
-  diagramDict.yOrderLookup = yOrder
+  // Drop any tables that were collected but
+  // lack fields and joins 
+  Object.keys(diagramDict.tableData).forEach(key => diagramDict.tableData[key] === undefined && delete diagramDict.tableData[key])
```
Appreciate the comments, they _help_ but this file in particular is pretty over-dense. consider multiple files, consider docstrings for all top level functions, and consider breaking some of these functions up into smaller functions that only do one thing.

RichardCzechowski reviewed (commented):
Yuge PR. I left some comments on a variety of ways this could be improved structurally. Some of it you may wish to save for follow up tech-debt PRs. Up to you.
In many cases I saw the same pattern used in multiple places, in those instances, I tried to just comment on one of them and indicate to find others yourself.

noahamac commented:
Yuge PR for sure, regretfully. I've addressed some low hanging items, and added tickets for later refactoring. We are staying in Beta, and are wanting to release this as v1.0.2 as it includes some bug fixes. 

@RichardCzechowski do you feel comfortable approving as an Open Beta patch release? 

Regarding the size of follow up PRs, 
are larger, more infrequent PRs, 
or smaller, potentially duplicative, more frequent PRs,
more respectful of your time? I'm sure a balance exists, and I feel like these need to be smaller and more focused. 

Thanks so much!

RichardCzechowski reviewed (approved):
Approving with the caveat that I'm not familiar with this codebase/project on any meaningful level. Additional feedback given via chat on how to improve the review experience and additional changes to be followed up on.
Thanks for the follow up tickets @noahamac
